### ISSUE TITLE
[FEATURE] Standardization of access to the classifiers pool in meta

### ISSUE THREAD

**Issue by luccaportes** [scikit-multiflow/scikit-multiflow#141]
Hi, in the [base ensemble class](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/base.py) in scikit-learn, they have the ```__iter__``` method overridden and the pool of classifiers stored in the variable ```self.estimators_```, so it is possible to do something like this:

```
pool = RandomForestClassifier()
for tree in pool:
    print(tree)
```
I find this very useful when working with research implementations of new methods that envolves dealing with attributes of each particular classifier in the ensemble.

Would something like this be possible in the scikit-multiflow?

I see that many of the implemented meta classifiers use self.ensemble to store the pool, but that is not true for all of them, such as [AccuracyWeightedEnsemble](https://github.com/scikit-multiflow/scikit-multiflow/blob/master/src/skmultiflow/meta/accuracy_weighted_ensemble.py), which uses ```self.models_pool```.

The easiest way, with fewer changes in the base code I can think, is overriding the ```__iter__``` method in the [MetaEstimatorMixin](https://github.com/scikit-multiflow/scikit-multiflow/blob/master/src/skmultiflow/core/base.py) to work with ```self.ensemble``` and also override in each ensemble implementation that uses a different variable name. So the new methods implemented would have to use ```self.ensemble``` or override ```__iter__```. 

This is not what scikit-learn does, they use the base ensemble class as an intermediate class between the implementations and the MetaEstimatorMixin.

Would this solution or any other solution to this be accepted in a PR?

**Comment by jacobmontiel** [comment]
Hello @luccaportes 

>Would something like this be possible in the scikit-multiflow?

Yes! And I agree with you on the potential of this functionality. I also like the idea of having a standard notation across for the ensembles. As you mention variable names are not always the same and this makes it difficult to maintain. I also like the generic approach to 'validate' and 'make' the ensemble.

>Would this solution or any other solution to this be accepted in a PR?

Yes! My suggestion is that the PR should focus on the implementation of the `BaseEnsemble` class + the standardization of the ensemble methods.

Optionally, we could sub-divide the PR into base class implementation + standardization given the number of methods available.

Comments @hmgomes @smastelini @AlbandeCrevoisier ?

Finally, could you elaborate on this:

>This is not what scikit-learn does, they use the base ensemble class as an intermediate class between the implementations and the MetaEstimatorMixin.

**Comment by AlbandeCrevoisier** [comment]
I agree that BaseEnsemble would be beneficial to the code base. Besides, I don't think merely extending MetaEstimatorMixin would be a good idea, meta estimators aren't necessarily ensembles. I'd say durably improving the code base prevails on minimising  changes — as well as maintaining the sklearn way of doing things.
I've been struggling with extending ARF, but I now have a 98% complete version that I can put in a PR for reference. BaseEnsemble would impact it — by making it easier to manage, actually — so I could work on it first, then adapt ARF & other ensemble methods as required? What do you think?

**Comment by smastelini** [comment]
Hello @luccaportes, @jacobmontiel, @AlbandeCrevoisier, and @hmgomes,

I think that creating a `BaseEnsemble` class would be a great addition to skmultiflow. I also agree with @AlbandeCrevoisier in the matter a meta estimator is not necessarily an ensemble. Hence, IMO, the best alternative would be creating an intermediate BaseEnsemble class, _a la_ sklearn. :D

> Yes! My suggestion is that the PR should focus on the implementation of the BaseEnsemble class + the standardization of the ensemble methods.

I believe this would be the fastest way to implement all the changes.

**Comment by luccaportes** [comment]
Thank you @jacobmontiel, @smastelini and @AlbandeCrevoisier

> Finally, could you elaborate on this:

I just meant that the scikit-learn BaseEnsemble extends the MetaEstimatorMixin.

So, just to be clear, the proposed approach would be something like this:

![scikit_alternative](https://user-images.githubusercontent.com/14240052/65618912-8a01b680-df95-11e9-8151-26be4076a4f0.png)

Is that right?

I was going to offer to implement whichever alternative that would outcome from this issue, but since @AlbandeCrevoisier is already working in something related, is there any way I can help?

**Comment by jacobmontiel** [comment]
+1 for this.

My comments:

* `BaseEnsemble` would inherit from both `MetaEstimatorMixin` and `BaseEstimator`.
* For compatibility with `scikit-learn` we should use the same variable name `ensemble_`
* For completeness, we can also add the `__len__` and `__getitem__ ` methods.

**Comment by AlbandeCrevoisier** [comment]
@luccaportes I didn't know whether you were willing to implement it, so was offering to do it in case you were merely requiring the feature. Your help would be most welcome! Besides, you seem more familiar with sklearn's implementation - I'm still new to this project.
I think you could implement BaseEnsemble, & then if you want I could help you with converting existing methods to use it? What do you think?

**Comment by AlbandeCrevoisier** [comment]
@luccaportes I didn't know whether you were willing to implement it, so was offering to do it in case you were merely requiring the feature. Your help would be most welcome! Besides, you seem more familiar with sklearn's implementation - I'm still new to this project.
I think you could implement BaseEnsemble, & then if you want I could help with converting existing methods to use it? What do you think?

**Comment by luccaportes** [comment]
@AlbandeCrevoisier, that would be great. I will work on that and keep you guys updated then.

Thanks

**Comment by jacobmontiel** [comment]
I am then assigning the issue to you @luccaportes.
Let us know if we can help.
